### ISSUE TITLE
Stable build does not make git tag available during build

### ISSUE THREAD

**Issue by djhoese** [readthedocs/readthedocs.org#10715]
## Details

* Read the Docs project URL: https://readthedocs.org/projects/satpy/
* Build URL (if applicable): https://readthedocs.org/projects/satpy/builds/21839497/
* Read the Docs username (if applicable): https://readthedocs.org/profiles/davidh-ssec/

## Expected Result

The "stable" build of my project is accurately grabbing the commit SHA associated with the most recent tag. I expect that during the build process there is enough git context for git commands to determine the name of the tag that is currently checked out. This is used heavily by dynamic version software (ex. `setuptools-scm`,  `versioneer`,  etc).

## Actual Result

During the `git checkout` process of getting the repository it does not actually have any of the tag information. So when my python package calls into `setuptools-scm` to determine the version number it uses a fallback (nonsense) version  number.

CC @gerritholl who identified the issue in our documentation.

If I run the same git commands as in the linked build above locally and then run setuptools-scm manually I get:

```bash
$ python -m setuptools_scm
/home/davidh/miniconda3/envs/satpy_py311_2/lib/python3.11/site-packages/setuptools_scm/git.py:135: UserWarning: "/tmp/rtd_test" is shallow and may cause errors
  warnings.warn(f'"{wd.path}" is shallow and may cause errors')
0.1.dev10921+g18288df
```

If I instead checkout all commits (just as a test), then checkout the specific SHA, then run it I get:

```bash
$ python -m setuptools_scm
/home/davidh/miniconda3/envs/satpy_py311_2/lib/python3.11/site-packages/setuptools_scm/git.py:135: UserWarning: "/tmp/rtd_test" is shallow and may cause errors
  warnings.warn(f'"{wd.path}" is shallow and may cause errors')
0.43.0
```

Am I maybe breaking something with my `.readthedocs.yaml`?

**Comment by humitos** [comment]
I'm not sure if this solves exactly what you are reporting, but have you read https://docs.readthedocs.io/en/latest/build-customization.html#avoid-having-a-dirty-git-index?

**Comment by djhoese** [comment]
@humitos Good guess and that may be needed in the future as I notice that the version number in our actual stable build is different from produced locally. On the RTD build:

```
Satpy 0.1.dev1407+g18288df.d20230807 documentation
```

And after re-reading the setuptools-scm README we can see the different variations of the version number:

https://github.com/pypa/setuptools_scm/#default-versioning-scheme

The above version would be the "distance and not clean". So git/setuptools-scm seems to be finding a "0.1" tag somewhere and detecting that our current commit is 1407 ahead of that and then it has the date of that commit. But it doesn't have any regular git tag information otherwise. I'll triple check on the 0.1 git tag and see where that is coming from.

Any other guesses? Has anything changed in the last ~6 months with how RTD gets code with git?

**Comment by humitos** [comment]
> From those two different builds, I noticed that the build from 2 months ago uses the _exact same commit_ than the build from an hour ago 🤷🏼

@stsewd I think you had a similar issue some time ago and open a PR to fix this and ran a small migration code (I think it was this one: https://github.com/readthedocs/readthedocs.org/pull/10606). Do you think this could be related? Maybe we missed some of the affected projects?

**Comment by stsewd** [comment]
@humitos the commit looks correct https://github.com/pytroll/satpy/commits/v0.43.0

![Screenshot 2023-09-13 at 19-06-52 Versions Read the Docs](https://github.com/readthedocs/readthedocs.org/assets/4975310/d32b4c82-7ba4-46cd-8cdc-6e9803ef63c8)

**Comment by djhoese** [comment]
Yes, the `v0.43.0` version that I activated specifically due to this issue does succeed with a realistic version (dirty, but that's a different issue), but that's because it explicitly fetches the tags with:

```
 git fetch origin --force --prune --prune-tags --depth 50 refs/tags/v0.43.0:refs/tags/v0.43.0 
```

Whereas the `stable` build only pulls the single commit:

```
 git fetch origin --force --prune --prune-tags --depth 50 18288df8dc943fc07322f47c435c08113abb0b8a 
```

**Comment by AntoineD** [comment]
We are now facing the same issue for this [doc](https://gemseo.readthedocs.io/en/stable/).

As already noticed above, the git commands have changed and now the tags are absent which causes tools like setuptools-scm to no longer be able to determine the version.

 It seems that the following git command
```
 git fetch origin --force --prune --prune-tags --depth 50 18288df8dc943fc07322f47c435c08113abb0b8a 
```
could be replaced by
```
 git fetch origin --force --prune --tags --depth 50 18288df8dc943fc07322f47c435c08113abb0b8a 
```
and the tags are back.

**Comment by humitos** [comment]
This is something that users can achieve with `build.jobs.post_checkout` to fetch the tags since it only affects projects using setuptools-scm.

**Comment by djhoese** [comment]
@humitos This does *not* only affect projects using setuptools-scm. It affects any project using git tags for versioning. Another example would be projects using versioneer or hatch-vcs. Anything that looks at the git tags won't find them.

And just because a tag and the "stable" version point to the same commit doesn't mean that RTD runs the same commands to get the git repository. It treats them as different things which I don't disagree with, but yeah git tags/history should almost always be checked out/fetched when getting a git repository.

**Comment by djhoese** [comment]
Here's a project of mine using `versioneer` with a public "stable" version of `1.27.1+0.g12892b8.dirty`:

https://pyresample.readthedocs.io/en/stable/

So part of that is just the RTD change to modify the git repository (making it "dirty") which I'm still not sure I understand. Would RTD be open to a pull request to make RTD work *without* modifying the git repository?

**Comment by humitos** [comment]
> git tags/history should almost always be checked out/fetched when getting a git repository.

I think this is what we want to avoid on the general use case because it penalizes all the projects by downloading more data than the one they need to build. I'd say that projects _requiring tags/history_ to be available for usage should use `build.jobs.post_checkout` to fetch the data they need. It's pretty simple:

```yaml
version: 2

build:
  jobs:
    post_checkout:
      - git fetch --tags
```

**Comment by stsewd** [comment]
This is also kind of problem with stable storing just the commit instead of the tag name, this is because we store "stable" as the name instead of the tag (similar situation happens with latest...).

**Comment by stsewd** [comment]
> This is also kind of problem with stable storing just the commit instead of the tag name, this is because we store "stable" as the name instead of the tag (similar situation happens with latest...).

In https://github.com/readthedocs/readthedocs.org/pull/10738 we fixed this situation in latest by tracking the name of the tag as the identifier, I think we can do the same for stable. But we should check if there are other parts of our code that depend on it being a commit (would be surprised if this was the case TBH).

**Comment by pllim** [comment]
I think I just hit a similar problem over at astropy: https://github.com/astropy/astropy/pull/15954
